scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -797,10 +809,187 @@ cncliPTsendslots() {
+  echo "Processing next epoch: ${next_epoch}"
+  stake_param_next="--active-stake ${active_stake_mark} --pool-stake ${pool_stake_mark}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --ledger-set next ${stake_param_next} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+}
+
+runPreviousEpochs() {
+  [[ -z ${KOIOS_API} ]] && return 1
+
+  if ! pool_hist=$(curl -sSL -f "${KOIOS_API}/pool_history?_pool_bech32=${POOL_ID_BECH32}&_epoch_no=${epoch}" 2>&1); then
+    echo "ERROR: Koios pool_stake_snapshot history query failed."
+    return 1
+  fi
+
+  if ! epoch_hist=$(curl -sSL -f "${KOIOS_API}/epoch_info?_epoch_no=${epoch}" 2>&1); then
+    echo "ERROR: Koios epoch_stake_snapshot history query failed."
+    return 1
+  fi
+
+  pool_stake_hist=$(jq -r '.[].active_stake' <<< "${pool_hist}")
+  active_stake_hist=$(jq -r '.[].active_stake' <<< "${epoch_hist}")
+
+  echo "Processing previous epoch: ${epoch}"
+  stake_param_prev="--active-stake ${active_stake_hist} --pool-stake ${pool_stake_hist}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${epoch}" --ledger-set prev ${stake_param_prev} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+
+  return 0
+}
+
+processAllEpochs() {
+  getCurrNextEpoch
+  IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+  for epoch in "${epochs_array[@]}"; do
+    if ! getConsensus "${epoch}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${epoch}."; return 1; fi
+    if [[ "$epoch" == "$curr_epoch" ]]; then
+      runCurrentEpoch
+    elif [[ "$epoch" == "$next_epoch" ]]; then
+      runNextEpoch
+    else
+      runPreviousEpochs
+    fi
+  done
+
+  id=1
+  while IFS= read -r row; do
+    echo "$id,$row" >> "$csvfile"
+    ((id++))
+  done < "$tmpcsv"
+
+  sqlite3 "$BLOCKLOG_DB" <<EOF
+DELETE FROM epochdata;
+VACUUM;
+.mode csv
+.import '$csvfile' epochdata
+REINDEX epochdata;
+EOF
+
+  row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata;")
+  echo "$row_count rows have been loaded into epochdata table in blocklog db"
+  echo "~ CNCLI epochdata table load completed ~"
+
+  rm $csvfile $tmpcsv
+}
+
+processSingleEpoch() {
+   getCurrNextEpoch
+   IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+   unset matched
+   for epoch in "${epochs_array[@]}"; do
+     [[ ${epoch} = "$1" ]] && matched=true && break
+   done
+   if [[ -z ${matched} ]]; then
+     echo -e "No slots found in blocklog table for epoch ${EPOCH}.\n"; return 1
+   fi
+   if ! getConsensus "${1}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${1}."; return 1; fi
+   if [[ "$1" == "$curr_epoch" ]]; then
+      runCurrentEpoch
+   elif [[ "$1" == "$next_epoch" ]]; then
+      runNextEpoch
+   else
+      runPreviousEpochs
+   fi
```

Comment:
Please pass $1 to all three functions runCurrentEpoch, runNextEpoch and runPreviousEpochs as param. 
Then as the first thing in each of these functions, do `local epoch="$1"`. Currently `${epoch}` is used in these functions but I dont think its properly set. 

Also do the same in `processAllEpochs()`, pass `${epoch}` there to all three functions. It might work anyway in some cases but this will make it much clearer when reading code.

Reply:
Hello @Scitz0 

thanks for all your feedbacks, regarding:

>Please pass $1 to all three functions runCurrentEpoch, runNextEpoch and runPreviousEpochs as param.....

I added your solution for runPreviousEpochs only.

runCurrentEpoch uses "${curr_epoch}" from getKoiosData and runNextEpoch doesn't need the epoch parameter.

Please let me know if it is really needed also for runCurrentEpoch and runNextEpoch.

Reply:
See new comment regarding --ledger-set.

Reply:
I will mark this conversation as resolved, lets continue discussion in new comment.